base/notify.go

```diff
@@ -0,0 +1,247 @@
+}
+
+// NotifyPayloadTemplateValues contains the report as well as the report in some other formats
+type NotifyPayloadTemplateValues struct {
+	// Report is the report
+	Report Report `json:"report" yaml:"report"`
+
+	// JSONStringReport is the report but marshalled into JSON and stringified
+	JSONStringReport string `json:"JSONStringReport" yaml:"JSONStringReport"`
+
+	// EscapedJSONStringReport is the report but marshalled into JSON and stringified and with tabs and new lines escaped
+	EscapedJSONStringReport string `json:"escapedJSONStringReport" yaml:"escapedJSONStringReport"`
+}
+
+// getPayloadTemplateValues gets the values for the payload tempalte
+func getPayloadTemplateValues(exp *Experiment) (*NotifyPayloadTemplateValues, error) {
+	report := Report{
+		// Group: exp.driver.Group
+		// SLOs:           slos,
+		// Metrics:        metrics,
+
+		TimeStamp:         time.Now().String(),
+		Completed:         exp.Completed(),
+		NoTaskFailures:    exp.NoFailure(),
+		NumTasks:          len(exp.Spec),
+		NumCompletedTasks: exp.Result.NumCompletedTasks,
+		NumLoops:          exp.Result.NumLoops,
+		Experiment:        exp,
+	}
+
+	marshalledReport, err := json.MarshalIndent(report, "", "  ")
+	if err != nil {
+		log.Logger.Error("could not JSON marshall report")
+		return nil, err
+	}
+
+	stringReport := string(marshalledReport)
+
+	// escape double quotes, tabs, and new lines
+	escapedReport := strings.Replace(stringReport, "\"", "\\\"", -1)
+	escapedReport = strings.Replace(escapedReport, "\n", "\\n", -1)
+	escapedReport = strings.Replace(escapedReport, "\t", "\\t", -1)
+
+	return &NotifyPayloadTemplateValues{
+		Report:                  report,
+		JSONStringReport:        stringReport,
+		EscapedJSONStringReport: escapedReport,
+	}, nil
+}
+
+// getPayload fetches the payload template from the PayloadTemplateURL and
+// executes it with values from getPayloadTemplateValues()
+func (t *notifyTask) getPayload(exp *Experiment) (string, error) {
+	if t.With.PayloadTemplateURL != "" {
+		template, err := getProviderTemplate(t.With.PayloadTemplateURL)
+		if err != nil {
+			log.Logger.Error("could not get payload template")
+			return "", err
+		}
+
+		values, err := getPayloadTemplateValues(exp)
+		if err != nil {
+			log.Logger.Error("could not get payload template values")
+			return "", err
+		}
+
+		// get the metrics spec
+		var buf bytes.Buffer
+		err = template.Execute(&buf, values)
+		if err != nil {
+			log.Logger.Error("could not execute payload template")
+			return "", err
+		}
+
+		return buf.String(), nil
+	}
+
+	return "", nil
+}
+
+// initializeDefaults sets default values for the custom metrics task
+func (t *notifyTask) initializeDefaults() {
+	// set default HTTP method
+	if t.With.Method == "" {
+		if t.With.PayloadTemplateURL != "" {
+			t.With.Method = "POST"
+		} else {
+			t.With.Method = "GET"
+		}
+	}
+}
+
+// validate task inputs
+func (t *notifyTask) validateInputs() error {
+	return nil
+}
```

Comment:
I feel like the url and PayloadTemplateURL should be verified as existing. They are set in current helm templates but this validation should cover other possibilities.

Reply:
Okay, we can do that.

Reply:
`PayloadTemplateURL` is optional as far as the task is concerned -- isn't it? (i.e., notification without payload?). I agree we verify `url`. I also agree we can verify `payloadTemplateURL`s within the slack and GitHub task templates.